api/test/test_helper/base_test_class.gd

```diff
@@ -0,0 +1,85 @@
+# ##############################################################################
+#
+#	Copyright 2020 Google LLC
+#
+#	Licensed under the Apache License, Version 2.0 (the "License");
+#	you may not use this file except in compliance with the License.
+#	You may obtain a copy of the License at
+#
+#		https://www.apache.org/licenses/LICENSE-2.0
+#
+#	Unless required by applicable law or agreed to in writing, software
+#	distributed under the License is distributed on an "AS IS" BASIS,
+#	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#	See the License for the specific language governing permissions and
+#	limitations under the License.
+#
+# ##############################################################################
+#
+# Base test class that inherits from gut/test.gd class, and provides a set of
+# helper functions for test classes to use.
+#
+# ##############################################################################
+extends "res://addons/gut/test.gd"
+
+func assert_asset_location_eq_dict(asset_location: PlayAssetLocation, dict: Dictionary):
+	assert_eq(dict.size(), 3)
+	assert_eq(asset_location.get_offset(), dict["offset"])
+	assert_eq(asset_location.get_path(), dict["path"])
+	assert_eq(asset_location.get_size(), dict["size"])
+
+func assert_asset_pack_location_eq_dict(asset_pack_location: PlayAssetPackLocation, dict: Dictionary):
+	assert_eq(dict.size(), 3)
+	assert_eq(asset_pack_location.get_assets_path(), dict["assetsPath"])
+	assert_eq(asset_pack_location.get_storage_method(), dict["packStorageMethod"])
+	assert_eq(asset_pack_location.get_path(), dict["path"])
+
+func assert_asset_pack_state_eq_dict(asset_pack_state: PlayAssetPackState, dict: Dictionary):
+	assert_eq(dict.size(), 6)
+	assert_eq(asset_pack_state.get_name(), dict["name"])
+	assert_eq(asset_pack_state.get_status(), dict["status"])
+	assert_eq(asset_pack_state.get_error_code(), dict["errorCode"])
+	assert_eq(asset_pack_state.get_bytes_downloaded(), dict["bytesDownloaded"])
+	assert_eq(asset_pack_state.get_total_bytes_to_download(), dict["totalBytesToDownload"])
+	assert_eq(asset_pack_state.get_transfer_progress_percentage(), dict["transferProgressPercentage"])
+
+func assert_asset_pack_states_eq_dict(asset_pack_states: PlayAssetPackStates, dict: Dictionary):
+	assert_eq(dict.size(), 2)
+	assert_eq(asset_pack_states.get_total_bytes(), dict["totalBytes"])
+	# tests get_pack_states()
+	var pack_states: Dictionary = asset_pack_states.get_pack_states()
+	assert_eq(pack_states.size(), dict["packStates"].size())
+	for key in dict["packStates"].keys():
+		assert_true(key in pack_states)
+		assert_asset_pack_state_eq_dict(pack_states[key], dict["packStates"][key])
+		# tests get_pack_state()
+		assert_asset_pack_state_eq_dict(asset_pack_states.get_pack_state(key), dict["packStates"][key])
+
+func create_mock_asset_pack_states_dict() -> Dictionary:
+	var pack_1_dict = {
+		"name": "assetPack1", 
```

Comment:
Should we use const for these key strings? But I don't see a place where we can declare the const vars currently.

Reply:
Could we put the const vars in the classes their associated with? i.e AssetPackState has:
````
const _bytes_downloaded_key = "bytesDownloaded"
const _total_bytes_downloaded_key = "totalBytesDownloaded"
etc.
````

Reply:
We should also consider making the dictionary values constants in this class where they can referenced by other tests. (i.e. line 62 could be changes from `"bytesDownloaded": 562` -> `"bytesDownloaded": pack_1_downloaded_halfway`)

Reply:
I tried declaring const variables inside classes such as PlayAssetLocation such as ```const _asset_pack_location_key = "offset". However in this case developers are able to see  and access the ``` __asset_pack_location_key ``` variable when then are handling a PlayAssetLocation object. This is because GDScript has no real implementation for private variables. The notion of using underscores is only a coding convention laid out by GDScript style guide that tells developer you should treat this variable as if it is private.

In order to solve this limitation, I decided to declare the const variables inside base_test_class.gd, so that all the test cases can reference these variables.

Reply:
Right now since the assertions are made by helper functions like [here](https://github.com/Arivval/Godot-Play-Asset-Delivery/blob/febfdf26cba8277f0c8df19cb8e2fcd8a9ad2aad/api/test/model/test_play_asset_pack_states.gd#L23) and [here](https://github.com/Arivval/Godot-Play-Asset-Delivery/blob/febfdf26cba8277f0c8df19cb8e2fcd8a9ad2aad/api/test/test_helper/base_test_class.gd#L72). Test cases don't need to reference the dictionary values again.